### DISCUSSION TITLE
Problems with running keycloak-js in backend

### DISCUSSION THREAD

**Discussion by camaeel** [keycloak/keycloak#8543] [Q&A]
I am using keycloak-admin (https://www.npmjs.com/package/keycloak-admin). And when I try to build & run my backend I get an error (in the runtime): 
```
root@743f5de4a1cf:/app# node packages/backend
/app/packages/backend/dist/index.cjs.js:25043
})(window, function (sha256_imported, base64js_imported) {
   ^

ReferenceError: window is not defined
    at /app/packages/backend/dist/index.cjs.js:25043:4
    at createCommonjsModule (/app/packages/backend/dist/index.cjs.js:90:9)
    at Object.<anonymous> (/app/packages/backend/dist/index.cjs.js:25036:16)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
```

This is caused by usage of window variable here: https://github.com/keycloak/keycloak/blob/443bd4a1baa229cf674cb2b617df8838fc92d22d/adapters/oidc/js/src/main/resources/keycloak.js#L63
```
})(window, function (sha256_imported, base64js_imported) {
```

If window is substituted with {} or commonjsGlobal defined like this: 
```
var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};
```
the process starts correctly.

**Comment by stianst** [comment]
@jonkoops FIY

**Comment by jonkoops** [comment]
The ` keycloak-admin` package has been deprecated and replaced with [`@keycloak/keycloak-admin-client`](https://www.npmjs.com/package/@keycloak/keycloak-admin-client). Could you install this dependency instead and report back if the same issue occurs?

**Reply by camaeel** [reply]
You are right. I pointed wrong package. I am using:     `"@keycloak/keycloak-admin-client": "^15.0.2"`
So the whole issue is still valid.

**Reply by jonkoops** [reply]
Thanks for checking. Under normal circumstances `keycloak-js` should never be included in the actual Admin Client if on Node, this should happen only in the browser.

I think that you might be using some kind of bundler which is transforming [this code](https://github.com/keycloak/keycloak-nodejs-admin-client/blob/master/src/client.ts#L95-L99) so that it is always included. If that is the case I would refer to the documentation of that specific tool to prevent this from happening.

I tightened the check for a browser environment in the 'next' version, perhaps you can try installing that to see if the result is still the same?

```bash
npm install @keycloak/keycloak-admin-client@next
```

**Reply by camaeel** [reply]
Sorry to say, but this dev version still generates such error.

**Reply by jonkoops** [reply]
Then I think this is really a problem with whatever bundler is being used to build your application together. I'd recommend looking there to see if the problem can be resolved.
